### ISSUE TITLE
Resolver can explode in solution size or backtrack options causing it to take too long to finish

### ISSUE THREAD

**Issue by ericmj** [hexpm/hex#630]
See https://github.com/elixir-lang/elixir/issues/8291.

Likely culprit commit: https://github.com/hexpm/hex/commit/c47e73aa35f2f3d1cb82a786ddcc8432398e96c2

/cc @dch

**Comment by michalmuskala** [comment]
I wonder if decreasing the general complexity here could help. Leveraging the ets match specs in `Version` we could easily create a function that matches multiple versions against multiple requirements in one go. I believe this should speed things up in general and possibly lower memory use because of less temporary allocations for version checking.

Also, right now we build the list of all solutions eagerly to sort them and find the best one. What if we did it lazily - it should be enough to only remember the best solution so far while traversing the requirements.

**Comment by ericmj** [comment]
Thanks @michalmuskala, I will investigate and measure your suggestions. But I am not sure it will do much of a difference because it looks like we need order of magnitudes performance improvements to fix this issue which rather suggests that it's an issue with the overall algorithm we use.

I will start with reverting the offending commit. After that I will try Michal's suggestions and profile to find low-hanging fruit. But I think we need to investigate using another algorithm, such as SAT solver or something more specific such as PubGrub [1]. The problem is that things such as optional dependencies and sub-tree overrides makes building the constraints very complicated.

[1] https://medium.com/@nex3/pubgrub-2fb6470504f

**Comment by michalmuskala** [comment]
Given the original report contained information about using 3GB of memory, I believe lowering the memory use might speed up things significantly by avoiding GCs. But I agree there's probably no better way than to profile. Switching an algorithm sounds like a major change, though.

**Comment by michalmuskala** [comment]
Another thing to try could be compiling the resolution modules with HiPE on the fly on first use of hex. But it needs some careful checking if it actually speeds things up and it might need some code restructuring as well - especially in using `:lists` instead of `Enum`, etc.

**Comment by dch** [comment]
FWIW I still see this issue (under 1.9.3 & OTP 22.1.5)  consumption of up to 30 Gb RAM under some circumstances. In particular I think the gen_rmq, amqp, amqp_client & rabbit_common modules all have tagged dependencies to work around issues spotted at different layers. I'll PR all of those and I believe this will settle down to a more reasonoble level.

Unless more people report this issue, I suggest closing this and hopefully it can be found by future people.

TLDR check to see if your deps and all subsequent deps are too tightly pinned, or even being overridden & that should help enormously.

**Comment by ericmj** [comment]
Thanks for letting us know. As far as I know only the amqp family of packages triggers this issue and the maintainers are already aware of the problem. I am keeping this issue open so we remember to verify the new resolver against it.

**Comment by ericmj** [comment]
The current plan is to replace the resolver with an algorithm that can more aggressively trim the possible solution space when backtracking. When that is done we can revisit these issues to see if they have been resolved.
